### PR TITLE
Modify REST API to allow operation execution on selected instances

### PR DESCRIPTION
**Pull request by adidanes**
# Pull Request description

As: An app manager
I want to: execute a workflow or a custom command on a specific instance or set of instances, while currently it is executed on all instances
So that: I can have a better control on runtime operations

AC1: works for both workflows and custom commands

## Description of the change

Make possible to execute **custom commands** on a selection of the target node's instances .

In the case of a **custom workflow**, it may contain several nodes (the workflow's steps may have different target nodes). So, this feature allows users to make a selection of instances for some nodes of a custom worklow.

The selected instances are identified by their ids that can be found using the `yorc deployments info -d` command , or by consulting the deployment's runtime view in the Alien4Cloud GUI.

As before, if no instances are selected for the target node(s), the execution takes place on all the running instances.

### What I did
Updated the REST API to support instances selection on target nodes for custom commands and custom workflows:

- POST /deployments/<deployment_id>/custom

- POST /deployments/<deployment_id>/workflows/<workflow_name>

In both cases the instances selection can be provided in the request body.

Updated the following Yorc CLI commands:

- yorc d custom

- yorc d wf exec 

In both cases the instances selection can be provided using the --data flag.

Updated the Yorc CLI commands documentation.

### How to verify it

The bellow commands were executed on a deployment made using attached CSAR

#### REST API example tests with curl

curl -X POST -v --noproxy localhost,127.0.0.1 --header 'Accept: application/json' --header 'Content-Type: application/json' 'localhost:8800/deployments/test-consul-simple-Environment/custom' --data '{ "node": "ConsulServer", "name": "maintenance_on", "interface": "custom", "instances": [ "0" ]}'

curl -X POST -v --noproxy localhost,127.0.0.1 --header 'Accept: application/json' --header 'Content-Type: application/json' 'localhost:8800/deployments/test-consul-simple-Environment/workflows/ServerMaintenanceOff' --data '{ "nodesinstances": [{ "name": "ConsulServer", "instances": [ "0" ] }] }'

#### CLI commans examples

yorc d custom test-consul-simple-Environment --data '{ "node": "ConsulServer", "name": "maintenance_off", "interface": "custom", "instances": [ "0" ]}'

yorc d wf exec test-consul-simple-Environment -w ServerMaintenanceOff --data '{ "nodesinstances": [{ "name": "ConsulServer", "instances": [ "1" ] }] }'

## Applicable Issues

https://github.com/ystia/yorc/issues/107

### CODE REVIEW

**Comment by loicalbertin** [comment]
Don't forget to make it work with the CLI.

**Comment by adidanes** [comment]
CSAR allowing to deploy app for test custom commands and workflows
[dep_csar.zip](https://github.com/ystia/yorc/files/3138824/dep_csar.zip)

File: rest/http_api.md [hunk 1]
```diff
@@ -616,7 +624,20 @@ Request body:
...
 
-If omitted interface defaults to `custom`.
+Request body allowing to execute command on some selected node instances:
+
+```json
+{
+    "node": "NodeName",
+    "name": "Custom_Command_Name",
+    "interface": "fully.qualified.interface.name",
+    "instances": "[ "0", "1" ]",
```
Comment (stebenoist) [inline rest/http_api.md]: You should use single quotes or escape double quotes inside first ones as "[ \"0\", \"1\" ]"

File: rest/http_api.md [hunk 2]
```diff
@@ -655,11 +676,29 @@ This endpoint will failed with an error "400 Bad Request" if:
...
 `POST /deployments/<deployment_id>/workflows/<workflow_name>[?continueOnError]`
 
+Request body allowing to execute a workflow's steps on selected node instances :
+
+```json
+{
+    "nodeinstances": [
+       {
+         "node": "Node_Name",
+         "instances": "[ "0", "2" ]"
```
Comment (stebenoist) [inline rest/http_api.md]: same as before

File: rest/dep_workflow.go [hunk 3]
```diff
@@ -61,7 +64,21 @@ func (s *Server) newWorkflowHandler(w http.ResponseWriter, r *http.Request) {
...
 		data["continueOnError"] = strconv.FormatBool(false)
 	}
-
+	// Get instances selection if provided in the request body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Panic(err)
+	}
+	var wfRequest WorkflowRequest
+	if err = json.Unmarshal(body, &wfRequest); err != nil {
```
Comment (stebenoist) [inline rest/dep_workflow.go]: Why not checking if body content is not empty before json unmarshall ?

File: commands/deployments/workflows/dep_wf_exec.go [hunk 4]
```diff
@@ -74,8 +81,9 @@ func init() {
 			return nil
 		},
 	}
-	wfExecCmd.PersistentFlags().StringVarP(&workflowName, "workflow-name", "w", "", "The workflows name")
+	wfExecCmd.PersistentFlags().StringVarP(&workflowName, "workflow-name", "w", "", "The workflows name (mandatory)")
 	wfExecCmd.PersistentFlags().BoolVarP(&continueOnError, "continue-on-error", "", false, "By default if an error occurs in a step of a workflow then other running steps are cancelled and the workflow is stopped. This flag allows to continue to the next steps even if an error occurs.")
+	wfExecCmd.PersistentFlags().StringVarP(&jsonParam, "data", "d", "", "Provide the JSON format for the node instances selection")
```
Comment (stebenoist) [inline commands/deployments/workflows/dep_wf_exec.go]: You can mention an array of instances names is expected

File: rest/dep_custom.go [hunk 5]
```diff
@@ -63,9 +63,12 @@ func (s *Server) newCustomCommandHandler(w http.ResponseWriter, r *http.Request)
 	}
 
 	data := make(map[string]string)
-
-	// For now custom commands are for all instances
-	instances, err := deployments.GetNodeInstancesIds(s.consulClient.KV(), id, ccRequest.NodeName)
+	// Get selected instances
+	instances := ccRequest.Instances
```
Comment (stebenoist) [inline rest/dep_custom.go]: Maybe checking if instances exist can be useful ?

File: rest/dep_workflow.go [hunk 6]
```diff
@@ -61,7 +64,21 @@ func (s *Server) newWorkflowHandler(w http.ResponseWriter, r *http.Request) {
...
+		log.Debugf("Custom workflow %s request body error : %s (probably no body as no instances selected)", workflowName, err)
+	} else {
+		for _, nodeInstances := range wfRequest.NodesInstances {
+			nodeName := nodeInstances.NodeName
```
Comment (stebenoist) [inline rest/dep_workflow.go]: A check on node/instances ?

**Comment by stebenoist** [comment]
Tests are OK for custom commands and workflows.

File: doc/cli.rst [hunk 7]
```diff
@@ -244,6 +244,7 @@ Trigger a workflow on deployment <DeploymentId>.
      yorc deployments workflows execute <DeploymentId> [flags]
```
Comment (stebenoist) [inline doc/cli.rst]: an example could be useful for executing custom workflows as for custom commands to see the required json format

File: rest/dep_custom.go [hunk 8]
```diff
@@ -113,3 +139,27 @@ func (s *Server) getInputNameFromCustom(deploymentID, nodeName, interfaceName, c
...
+// If a provided instance does not exists, returns false and the inexistent instance name
+func (s *Server) checkInstances(deploymentID, nodeName string, provInstances []string) (bool, string) {
+	// Get known instances for the provided node
+	allInstances, err := deployments.GetNodeInstancesIds(s.consulClient.KV(), deploymentID, nodeName)
+	if err != nil {
+		log.Panic(err)
+	}
+	// Check that provided instances exist
+	for _, provInstance := range provInstances {
+		var found bool
```
Comment (stebenoist) [inline rest/dep_custom.go]: You can make it simpler here with collections.ContainsString

File: rest/http_api.md (hunk 1 above)
Comment (adidanes) [reply rest/http_api.md]: Fixed to 
"instances": [ "0", "1" ]

File: rest/dep_workflow.go (hunk 3 above)
Comment (adidanes) [reply rest/dep_workflow.go]: Took example on dep_custom.go 
It exists a function to check that a request has an empty body ?

File: rest/http_api.md [hunk 9]
```diff
@@ -211,7 +211,8 @@ Content-Type: application/json
 
 ### Get the deployment information about a given node instance <a name="instance-info"></a>
 
-Retrieve the node instance status and the list (as Atom links) of the attributes for this instance.
+Retrieve a node instance's status.
```
Comment (loicalbertin) [inline rest/http_api.md]:
```suggestion
Retrieve a node instance's status.
```

File: rest/http_api.md [hunk 10]
```diff
@@ -583,8 +584,9 @@ Content-Length: 0
 
 ### Resume a task <a name="task-resume"></a>
 
-Resume a task for a given deployment. The task should be in status "FAILED" to be resumed otherwise an HTTP 400
-(Bad request) error is returned.
+Resume a task for a given deployment.
```
Comment (loicalbertin) [inline rest/http_api.md]:
```suggestion
Resume a task for a given deployment.
```

File: rest/http_api.md [hunk 11]
```diff
@@ -597,12 +599,18 @@ Content-Length: 0
 
 ### Execute a custom command <a name="custom-cmd-exec"></a>
 
-Submit a custom command for a given deployment.
+Submit a custom command for a given deployment.
```
Comment (loicalbertin) [inline rest/http_api.md]:
```suggestion
Submit a custom command for a given deployment.
```

File: rest/http_api.md [hunk 12]
```diff
@@ -655,11 +676,29 @@ This endpoint will failed with an error "400 Bad Request" if:
 
 ### Execute a workflow <a name="workflow-exec"></a>
 
-Submit a custom workflow for a given deployment. By adding the optional 'continueOnError' url parameter to your request workflow will
-not stop at the first encountered error and will run to its end.
+Submit a custom workflow for a given deployment.
```
Comment (loicalbertin) [inline rest/http_api.md]:
```suggestion
Submit a custom workflow for a given deployment.
```

File: rest/dep_custom.go (hunk 8 above)
Comment (loicalbertin) [reply rest/dep_custom.go]: +1

**Pull request merged**